### ISSUE TITLE
ops.VideoReader infinite loop

### ISSUE THREAD

**Issue by km23630** [NVIDIA/DALI#1644]
I want to use NVIDIA Dali to inference.
previous video load (cv2.videocapture) has bottle neck 
so I want loading video using Dali.

I followed NVIDIA's video read tutorial
process was successful. 
but ops.VideoReader generate infinite batches.

original video (14 frame)
[■■■■■■■■■■■■■■]

current VideoReader output 
[[■■■■], [■■■■], [■■■■] … [■■■■], [■■■■]]

VideoReader output which I want (batch_size=4)
[[■■■■], [■■■■], [■■■■], [■■]]

I searched NVIDIA Document but can't found solution.
please solve my problem.
thank you!

**Comment by szalpal** [comment]
@km23630 ,
Could you paste some code? Preferably, how you implemented Dali Pipeline

**Comment by JanuszL** [comment]
One more comment - DALI's primary use case, for now, is training, so readers are designed as an infinite data source that just wraps up. You can pass the data size to DALI FW iterators (PyTorch, MXNet or PaddlePaddle one) so they will stop after traversing the whole data set.
In your case you should just stop iterating over the data when you know you have read enough samples, otherwise, you would read the same one over and over again.

**Comment by km23630** [comment]
`import time

import cv2
import math
import numpy as np

from nvidia.dali.pipeline import Pipeline
import nvidia.dali.ops as ops


class VideoPipe(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, data, shuffle):
        super(VideoPipe, self).__init__(batch_size, num_threads, device_id, seed=16)
        self.input = ops.VideoReader(device="gpu", filenames=data, sequence_length=sequence_length,
                                     shard_id=0, num_shards=1, pad_last_batch=True,
                                     random_shuffle=shuffle, initial_fill=initial_prefetch_size)

    def define_graph(self):
        output = self.input(name="Reader")
        return output


if __name__ == '__main__':
    batch_size = 4
    sequence_length = 1
    initial_prefetch_size = 16
    
    video_files = ['./video/test.mp4']
    cap = cv2.VideoCapture(video_files[0])
    
    total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
    n_iter = math.ceil(total_frames/batch_size)
    last_padding_frames = (n_iter * batch_size) - total_frames

    pipe = VideoPipe(batch_size=batch_size, num_threads=2, device_id=0, data=video_files, shuffle=False)
    pipe.build()
    
    ptime = time.time()
    for i in range(n_iter):
        pipe_out = pipe.run()
        sequences_out = list(pipe_out[0].as_cpu().as_array())
        if i == (n_iter - 1):
            sequences_out = sequences_out[:-last_padding_frames]
            [sequences_out.append(np.zeros_like(sequences_out[0])) for i in range(last_padding_frames)]
        print(np.shape(sequences_out))
      
    print("dali", time.time() - ptime, "msec")`


this is my code snippet

**Comment by JanuszL** [comment]
@km23630 - I see you specify on your own the number of batches you want to get from DALI `n_iter`. Could you be more specific about what kind of problem you are experiencing with the posted code?

**Comment by km23630** [comment]
@JanuszL 

yes this code works fine!
but I think this code is not elegant, so I asked if there is a better way.

like this
 `if i == (n_iter - 1):
            sequences_out = sequences_out[:-last_padding_frames]`

my problem was solved.
thanks :)

**Comment by JanuszL** [comment]
@km23630 - there is no better way now. Framework iterators do similar things under the hood, to be honest.
